### DISCUSSION TITLE
How to preload an image only when not in cache

### DISCUSSION THREAD

**Discussion by kk-atom** [coil-kt/coil#1835] [Q&A]
https://coil-kt.github.io/coil/getting_started/#preloading
```
val request = ImageRequest.Builder(context)
    .data("https://example.com/image.jpg")
    .build()
imageLoader.enqueue(request)
```
Code above always loads an image from network. How to preload an image only when not in disk cache?
How to check if image is cached?

**Comment by kk-atom** [comment]
This seems to do the job.
```
        val urlDiskCacheSnapshot = context.imageLoader.diskCache?.openSnapshot(url)
        val shouldPreload = urlDiskCacheSnapshot == null
        try {
            urlDiskCacheSnapshot?.close()
        } catch (_: Exception) {
        }
        if (shouldPreload) {
            val request = ImageRequest.Builder(context)
                .data(url)
                .build()
            context.imageLoader.enqueue(request)
        }
```

**Comment by DronPascal** [comment]
> This seems to do the job.
> 
> ```
>         val urlDiskCacheSnapshot = context.imageLoader.diskCache?.openSnapshot(url)
>         val shouldPreload = urlDiskCacheSnapshot == null
>         try {
>             urlDiskCacheSnapshot?.close()
>         } catch (_: Exception) {
>         }
>         if (shouldPreload) {
>             val request = ImageRequest.Builder(context)
>                 .data(url)
>                 .build()
>             context.imageLoader.enqueue(request)
>         }
> ```

HttpUriFetcher checks disk сache itself.
This
```
         val urlDiskCacheSnapshot = context.imageLoader.diskCache?.openSnapshot(url)
         val shouldPreload = urlDiskCacheSnapshot == null
         try {
                  urlDiskCacheSnapshot?.close()
         } catch (_: Exception) {
         }
```
doesn't needed

**Reply by kk-atom** [reply]
```
val request = ImageRequest.Builder(context)
    .data("https://example.com/image.jpg")
    .build()
imageLoader.enqueue(request)
```
1. Execute the code when https://example.com/image.jpg returns 200
2. Image is cached
3. Execute the code but now when https://example.com/image.jpg returns 403
4. Image is cleared from cache.
      
Following code uses cache
```
AsyncImage(
   model = "https://example.com/image.jpg",
   contentDescription = null,
)
```
1. Execute the code when https://example.com/image.jpg returns 200
2. Image is cached
3. Execute the code but now when https://example.com/image.jpg returns 403
4. Image from cache is used.
